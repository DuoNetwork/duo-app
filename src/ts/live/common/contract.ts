import infura from '../../../../../duo-admin/src/keys/infura.json';
import ContractUtil from '../../../../../duo-contract-util/src/contractUtil';
import * as CST from './constants';

const provider =
	(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) + '/' + infura.token;
const contract = new ContractUtil(window, '', provider, !__KOVAN__);
export default contract;
