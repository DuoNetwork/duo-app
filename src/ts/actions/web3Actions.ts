import * as CST from 'ts/common/constants';
import { VoidThunkAction } from 'ts/common/types';
import { web3Wrapper } from 'ts/common/wrappers';

export function accountUpdate(account: string) {
	return {
		type: CST.AC_WEB3_ACCOUNT,
		value: account
	};
}

export function getAccount(): VoidThunkAction {
	return async dispatch => dispatch(accountUpdate(await web3Wrapper.getCurrentAddress()));
}

export function networkUpdate(networkId: number) {
	return {
		type: CST.AC_WEB3_NETWORK,
		value: networkId
	};
}

export function getNetwork(): VoidThunkAction {
	return async dispatch => dispatch(networkUpdate(await web3Wrapper.getCurrentNetwork()));
}

export function gasPriceUpdate(gasPrice: number) {
	return {
		type: CST.AC_WEB3_GAS_PX,
		value: gasPrice
	};
}

export function getGasPrice(): VoidThunkAction {
	return async dispatch => dispatch(gasPriceUpdate(await web3Wrapper.getGasPrice()));
}

export function balanceUpdate(balance: number) {
	return {
		type: CST.AC_WEB3_BALACE,
		value: balance
	};
}

export function getBalance(): VoidThunkAction {
	return async (dispatch, getState) =>
		dispatch(balanceUpdate(await web3Wrapper.getEthBalance(getState().web3.account)));
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getAccount());
		dispatch(getGasPrice());
		dispatch(getNetwork());
		dispatch(getBalance());
	};
}
