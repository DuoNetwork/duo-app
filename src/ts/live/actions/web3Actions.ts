import * as CST from '../common/constants';
import contract from '../common/contract';
import { VoidThunkAction } from '../common/types';

export function accountUpdate(account: string) {
	return {
		type: CST.AC_ACCOUNT,
		value: account
	};
}

export function getAccount(): VoidThunkAction {
	return async dispatch => dispatch(accountUpdate(await contract.getCurrentAddress()));
}

export function networkUpdate(networkId: number) {
	return {
		type: CST.AC_NETWORK,
		value: networkId
	};
}

export function getNetwork(): VoidThunkAction {
	return async dispatch => dispatch(networkUpdate(await contract.getCurrentNetwork()));
}

export function gasPriceUpdate(gasPrice: number) {
	return {
		type: CST.AC_GAS_PX,
		value: gasPrice
	};
}

export function getGasPrice(): VoidThunkAction {
	return async dispatch => dispatch(gasPriceUpdate(await contract.getGasPrice()));
}

export function refresh(): VoidThunkAction {
	return dispatch => {
		dispatch(getAccount());
		dispatch(getGasPrice());
		dispatch(getNetwork());
	};
}
