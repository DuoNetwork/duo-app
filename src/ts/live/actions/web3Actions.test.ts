import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import contract from '../common/contract';
import * as web3Actions from './web3Actions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('accountUpdate', () => {
		expect(web3Actions.accountUpdate('test')).toMatchSnapshot();
	});

	test('getAccount', () => {
		const store: any = mockStore({});
		contract.getCurrentAddress = jest.fn(() => Promise.resolve('0x0'));
		store.dispatch(web3Actions.getAccount());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('networkUpdate', () => {
		expect(web3Actions.networkUpdate(123)).toMatchSnapshot();
	});

	test('getNetwork', () => {
		const store: any = mockStore({});
		contract.getCurrentNetwork = jest.fn(() => Promise.resolve(123));
		store.dispatch(web3Actions.getNetwork());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('gasPriceUpdate', () => {
		expect(web3Actions.gasPriceUpdate(123)).toMatchSnapshot();
	});

	test('getGasPrice', () => {
		const store: any = mockStore({});
		contract.getGasPrice = jest.fn(() => Promise.resolve(456));
		store.dispatch(web3Actions.getGasPrice());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
