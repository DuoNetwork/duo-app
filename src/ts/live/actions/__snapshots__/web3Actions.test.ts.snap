// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions accountUpdate 1`] = `
Object {
  "type": "account",
  "value": "test",
}
`;

exports[`actions gasPriceUpdate 1`] = `
Object {
  "type": "gasPrice",
  "value": 123,
}
`;

exports[`actions getAccount 1`] = `
Array [
  Object {
    "type": "account",
    "value": "0x0",
  },
]
`;

exports[`actions getGasPrice 1`] = `
Array [
  Object {
    "type": "gasPrice",
    "value": 456,
  },
]
`;

exports[`actions getNetwork 1`] = `
Array [
  Object {
    "type": "network",
    "value": 123,
  },
]
`;

exports[`actions networkUpdate 1`] = `
Object {
  "type": "network",
  "value": 123,
}
`;
